client/src/game/ui/selection/shapecontext.vue

```diff
@@ -19,85 +19,139 @@
 
+import { mapState } from "vuex";
+
 import ContextMenu from "@/core/components/contextmenu.vue";
 
+import { socket } from "@/game/api/socket";
+import { ServerLocation } from "@/game/comm/types/general";
 import { EventBus } from "@/game/event-bus";
 import { layerManager } from "@/game/layers/manager";
 import { Shape } from "@/game/shapes/shape";
 import { gameStore } from "@/game/store";
-import { inInitiative, initiativeStore } from "../initiative/store";
+import { cutShapes, deleteShapes, pasteShapes } from "../../shapes/utils";
+import { initiativeStore, inInitiative } from "../initiative/store";
 
 @Component({
     components: {
         ContextMenu,
     },
+    computed: {
+        ...mapState("game", ["locations", "assets", "notes"]),
+    },
 })
 export default class ShapeContext extends Vue {
     visible = false;
     x = 0;
     y = 0;
-    shape: Shape | null = null;
+    shapes: Shape[] | null = null;
     get activeLayer(): string {
         const layer = layerManager.getLayer();
         return layer === undefined ? "" : layer.name;
     }
-    open(event: MouseEvent, shape: Shape) {
+    open(event: MouseEvent, shapes: Shape[]) {
         this.visible = true;
         this.x = event.pageX;
         this.y = event.pageY;
-        this.shape = shape;
+        this.shapes = shapes;
         this.$nextTick(() => (<HTMLElement>this.$children[0].$el).focus());
     }
     close() {
         this.visible = false;
-        this.shape = null;
+        this.shapes = null;
     }
     getLayers() {
         return layerManager.layers.filter(l => l.selectable && (gameStore.IS_DM || l.playerEditable));
     }
     getActiveLayer() {
         return layerManager.getLayer();
     }
+    getCurrentLocation() {
+        return gameStore.locationName;
+    }
     getInitiativeWord() {
-        if (this.shape === null) return "";
-        return inInitiative(this.shape.uuid) ? "Show" : "Add";
+        if (this.shapes === null) return "";
+        if (this.shapes.length === 1) {
+            return inInitiative(this.shapes[0].uuid) ? "Show" : "Add";
+        } else {
+            return this.shapes.every(shape => inInitiative(shape.uuid)) ? "Show" : "Add all to";
+        }
+    }
+    hasSingleShape(): boolean {
+        return this.shapes !== null && this.shapes.length === 1;
     }
     setLayer(newLayer: string) {
-        if (this.shape === null) return;
-        this.shape.moveLayer(newLayer, true);
+        if (this.shapes === null) return;
+        this.shapes.forEach(shape => shape.moveLayer(newLayer, true));
+        this.close();
+    }
+    setLocation(newLocation: string) {
+        if (this.shapes === null) return;
+        const layer = layerManager.getLayer()!;
+        layer.selection = this.shapes;
+        cutShapes();
+        // Request change to other location
+        socket.emit("Location.Change", newLocation);
+        socket.once("Location.Set", (_data: Partial<ServerLocation>) => {
```

Comment:
Do you know if there is any guarantee that this is always ran after the main listener in events.ts ?  Depending on which one fires first, the lighting is updated with the new shapes or without.

Reply:
Not sure. It might be guaranteed that recieving events are recieved in order of subscription. I think the correct solution would be to first emit a "paste to location" event, then change locations. However, i'm not sure if emiting events guarantees that the server recieves them in the same order, or that the first event will be totally processe before running the second one.

Reply:
It's plausible that they are processed in subscription order yes.

If I remember correctly the messages are guaranteed to arrive in the correct order. (at least in the case where you send multiple messages to others; if multiple people were sending stuff to you then the latency would be the deciding factor)

Reply:
I assume the messages would be recieved the right order. I mean if the first message would be completelly processed before the second message starts to be processed.

Reply:
In any case, we could just send a message back from the server with "onPasteFinished" or similar. Now we have the problem of what happens if the server finishes the paste before the client sets the callback. Probably just seting the callback before doing the paste would work anyway.

Reply:
I just checked the node docs and they are in the order they are registered. so the .once() will in this case always be called after the main .on() in events.ts